### PR TITLE
Graphics refactor #1

### PR DESCRIPTION
**Pull request by kkaefer**
No functionality changes, just moving code around to reduce included code.

### CODE REVIEW

**Comment by pozdnyakov** [comment]
@kkaefer the changes look nice overall :+1:, however patch is quite big and it somehow affects the binary size. Would it be feasible for you to split it into several PRs (e.g. 1 PR per commit) so that we can review it quickly and see what exact change causes binary size increase?

**Comment by kkaefer** [comment]
@pozdnyakov already found the culprit; this PR should now reduce the size overall. I separated out individual commits as well as possible; does that work for you? Submitting individual PRs for every commit seems like a lot of unnecessary work.

File: src/mbgl/gl/enum.cpp [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+template <>
+gfx::ColorBlendEquationType Enum<gfx::ColorBlendEquationType>::from(const platform::GLint value) {
+    switch (value) {
+        case GL_FUNC_ADD: return gfx::ColorBlendEquationType::Add;
+        case GL_FUNC_SUBTRACT: return gfx::ColorBlendEquationType::Subtract;
+        case GL_FUNC_REVERSE_SUBTRACT: return gfx::ColorBlendEquationType::ReverseSubtract;
+    }
+    return {};
```
Comment (alexshalamov) [inline src/mbgl/gl/enum.cpp]: should this be handled somehow? assert or log error?

**Comment by alexshalamov** [comment]
> @pozdnyakov already found the culprit;

@kkaefer What was the reason? Would be nice to know, to avoid it in the future.

**Comment by kkaefer** [comment]
@alexshalamov I instantiated various `PaintPropertyBinder` classes with the `a_*` attribute structs, which created a separate implementation. I reverted to just instantiating them with the type these structs contain which dramatically reduced the number of instantiations.

File: src/mbgl/gl/enum.cpp (hunk 1 above)
Comment (kkaefer) [reply src/mbgl/gl/enum.cpp]: I don't think so; in that case it defaults to an undefined value of `gfx::ColorBlendEquationType` which will be okay for most use cases. Additionally, we're only using the OpenGL state getters for debug purposes; they're not used in production builds.

File: src/mbgl/gfx/color_mode.hpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    using BlendFunction = variant<
+        Replace,
+        Add,
+        Subtract,
+        ReverseSubtract>;
+
+    BlendFunction blendFunction;
+    Color blendColor;
+
+    struct Mask {
```
Comment (pozdnyakov) [inline src/mbgl/gfx/color_mode.hpp]: nit: would a bit mask be better here instead?

File: src/mbgl/gfx/color_mode.hpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    static ColorMode disabled() {
+       return { Replace{}, {}, { false, false, false, false } };
+    }
+
+    static ColorMode unblended() {
+       return { Replace{}, {}, { true, true, true, true } };
+    }
+
+    static ColorMode alphaBlended() {
+        return { Add{ ColorBlendFactorType::One, ColorBlendFactorType::OneMinusSrcAlpha }, {}, { true, true, true, true } };
```
Comment (pozdnyakov) [inline src/mbgl/gfx/color_mode.hpp]: Could we provide actual types, e.g. s/`{}`/`Color()` , to improve readability?

File: src/mbgl/gfx/color_mode.hpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    static ColorMode additive() {
+        return { Add{ ColorBlendFactorType::One, ColorBlendFactorType::One }, {}, { true, true, true, true } };
+    }
+};
+
+constexpr bool operator!=(const ColorMode::Mask& a, const ColorMode::Mask& b) {
```
Comment (pozdnyakov) [inline src/mbgl/gfx/color_mode.hpp]: if we had bit mask, that would be just comparison of two `unsigned` types

File: src/mbgl/gfx/attribute.hpp [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+#include <mbgl/util/type_list.hpp>
+#include <mbgl/util/indexed_tuple.hpp>
+
+#include <array>
+#include <cstddef>
+
+#define MBGL_DEFINE_ATTRIBUTE(type_, n_, name_)                                                    \
+    struct name_ {                                                                                 \
+        using Type = ::mbgl::gfx::AttributeType<type_, n_>;                                        \
+        static auto name() {                                                                       \
```
Comment (pozdnyakov) [inline src/mbgl/gfx/attribute.hpp]: constexpr ?

File: src/mbgl/gfx/color_mode.hpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    struct Mask {
+        bool r;
+        bool g;
+        bool b;
+        bool a;
+    };
+
+    Mask mask;
+
+    static ColorMode disabled() {
```
Comment (pozdnyakov) [inline src/mbgl/gfx/color_mode.hpp]: here and below (in other structs), could we make the "factory" methods `constexpr` too?

**Pull request merged**